From 1edb42cc2890df4b57dd207223f6a10bb009d7c0 Mon Sep 17 00:00:00 2001
From: elpaablo <donopwhile1@gmail.com>
Date: Sat, 20 Jun 2026 12:50:50 +0100
Subject: [PATCH 1/2] wm: ActivityStarter: Redirect proprietary OPlus file
 manager intents to DocumentsUI

The OPlus camera/gallery attempts to launch proprietary file managers
(oplus.intent.action.filemanager.BROWSER_FILE or
oppo.filemanager.intent.action.BROWSER_FILE) which are not present
on AOSP/AlphaDroid.

This intercepts those intents, maps them to the standard DocumentsUI
directory view, and elevates the calling UID/Package to SYSTEM to
bypass caller URI permission checks that would otherwise throw a
SecurityException from the Gallery app.
---
 .../android/server/wm/ActivityStarter.java    | 61 +++++++++++++++++++
 1 file changed, 61 insertions(+)

diff --git a/services/core/java/com/android/server/wm/ActivityStarter.java b/services/core/java/com/android/server/wm/ActivityStarter.java
index bec59b1147f3..c6850cec0c93 100644
--- a/services/core/java/com/android/server/wm/ActivityStarter.java
+++ b/services/core/java/com/android/server/wm/ActivityStarter.java
@@ -586,6 +586,67 @@ class ActivityStarter {
             ephemeralIntent = new Intent(intent);
             // Don't modify the client's object!
             intent = new Intent(intent);
+            if (intent.getAction() != null &&
+                    ("oplus.intent.action.filemanager.BROWSER_FILE".equals(intent.getAction())
+                    || "oppo.filemanager.intent.action.BROWSER_FILE".equals(intent.getAction()))) {
+                String currentDir = intent.getStringExtra("CurrentDir");
+                if (currentDir != null) {
+                    // Extract parent directory if the path points to a file
+                    int lastSlash = currentDir.lastIndexOf('/');
+                    if (lastSlash > 0 && currentDir.substring(lastSlash).contains(".")) {
+                        currentDir = currentDir.substring(0, lastSlash);
+                    }
+
+                    String docId = null;
+                    if (currentDir.startsWith("/storage/emulated/0")) {
+                        String relPath = currentDir.substring("/storage/emulated/0".length());
+                        if (relPath.startsWith("/")) {
+                            relPath = relPath.substring(1);
+                        }
+                        docId = "primary:" + relPath;
+                    } else if (currentDir.startsWith("/storage/")) {
+                        String sub = currentDir.substring("/storage/".length());
+                        int slashIdx = sub.indexOf('/');
+                        if (slashIdx > 0) {
+                            String uuid = sub.substring(0, slashIdx);
+                            String relPath = sub.substring(slashIdx + 1);
+                            docId = uuid + ":" + relPath;
+                        } else {
+                            docId = sub + ":";
+                        }
+                    } else if (currentDir.startsWith("/sdcard")) {
+                        String relPath = currentDir.substring("/sdcard".length());
+                        if (relPath.startsWith("/")) {
+                            relPath = relPath.substring(1);
+                        }
+                        docId = "primary:" + relPath;
+                    }
+
+                    if (docId != null) {
+                        try {
+                            String encodedDocId = java.net.URLEncoder.encode(docId, "UTF-8")
+                                    .replace("+", "%20");
+                            android.net.Uri uri = android.net.Uri.parse(
+                                    "content://com.android.externalstorage.documents/document/" + encodedDocId);
+                            intent.setAction(Intent.ACTION_VIEW);
+                            intent.setDataAndType(uri, "vnd.android.document/directory");
+                            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
+                            intent.setComponent(null);
+
+                            // Elevate privileges to system so we bypass caller URI permission checks
+                            callingUid = android.os.Process.SYSTEM_UID;
+                            callingPackage = "android";
+                            realCallingUid = android.os.Process.SYSTEM_UID;
+                            realCallingPid = android.os.Process.myPid();
+                            callingPid = android.os.Process.myPid();
+                            resolvedCallingUid = android.os.Process.SYSTEM_UID;
+                            resolvedCallingPackage = "android";
+                        } catch (java.io.UnsupportedEncodingException e) {
+                            // ignore
+                        }
+                    }
+                }
+            }
             if (intent.getComponent() != null
                     && !(Intent.ACTION_VIEW.equals(intent.getAction()) && intent.getData() == null)
                     && !Intent.ACTION_INSTALL_INSTANT_APP_PACKAGE.equals(intent.getAction())
-- 
2.34.1

